File: v1/backend/storage/repos/9f9d7705-da3e-45fb-b4fd-3b342766baf8/tuck-in-tales-frontend/src/models/character.ts
export interface CharacterRelationship {
  id: string;
  from_character_id: string;
  to_character_id: string;
  to_character_name?: string;
  to_character_avatar_url?: string;
  relationship_type: string;
  created_at?: string;
}

export interface AddRelationshipRequest {
  to_character_id: string;
  relationship_type: string;
}

// Base Character type (matching backend)
export interface Character {
  id: string;
  name: string;
  bio?: string | null;
  birth_date?: string | null;
  visual_description?: string | null;
  photo_paths?: string[];
  avatar_url?: string | null;
  created_at: string;
  family_id: string;
  relationships?: CharacterRelationship[];
}

// Type for displaying in lists (subset of Character)
export interface CharacterSummary {
  id: string;
  name: string;
  avatar_url?: string | null;
}

// Type for detailed view (could be same as Character or have extra fields)
export interface CharacterDetail extends Character {
  // Add any additional fields specific to detail view if needed
  // e.g., associated_memories?: Memory[];
}

// Type for creating a new character (input to API)
export interface CharacterCreate {
  name: string;
  bio?: string | null;
  birth_date?: string | null; // Add optional birth_date
  // Add other required fields for creation
}

// Type for updating an existing character (input to API, fields are optional)
export interface CharacterUpdate {
  name?: string;
  bio?: string | null;
  birth_date?: string | null; // Add optional birth_date
  // Add other updatable fields
} 

File: v1/backend/storage/repos/9f9d7705-da3e-45fb-b4fd-3b342766baf8/tuck-in-tales-backend/src/models/character.py
from uuid import UUID, uuid4
from datetime import datetime, date
from pydantic import BaseModel, Field, HttpUrl
from typing import Optional, List

class CharacterBase(BaseModel):
    name: str = Field(..., min_length=1, max_length=100)
    bio: Optional[str] = Field(None, max_length=2000)
    photo_paths: Optional[List[str]] = None
    avatar_url: str | None = None
    birth_date: Optional[date] = Field(None, alias='birthdate')
    visual_description: Optional[str] = None
    # family_id: UUID # Removed - Determined by auth

class CharacterCreate(CharacterBase):
    pass

# Model for updating an existing character (all fields optional)
class CharacterUpdate(CharacterBase):
    name: Optional[str] = Field(None, min_length=1, max_length=100)
    bio: Optional[str] = Field(None, max_length=2000)
    photo_paths: Optional[List[str]] = None
    avatar_url: str | None = None
    birth_date: Optional[date] = Field(None, alias='birthdate')
    visual_description: Optional[str] = None
    # family_id cannot be changed

class CharacterRelationship(BaseModel):
    id: UUID
    from_character_id: UUID
    to_character_id: UUID
    to_character_name: Optional[str] = None
    to_character_avatar_url: Optional[str] = None
    relationship_type: str
    created_at: Optional[datetime] = None

    class Config:
        from_attributes = True


class AddRelationshipRequest(BaseModel):
    to_character_id: UUID
    relationship_type: str = Field(..., min_length=1, max_length=100)


class UpdateRelationshipRequest(BaseModel):
    relationship_type: str = Field(..., min_length=1, max_length=100)


class Character(CharacterBase):
    id: UUID
    family_id: UUID
    created_at: datetime = Field(default_factory=datetime.utcnow)
    updated_at: Optional[datetime] = None
    relationships: Optional[List[CharacterRelationship]] = None

    class Config:
        from_attributes = True
        populate_by_name = True